206.ReverseLinkedList/step1.cpp

```diff
@@ -0,0 +1,36 @@
+/**
+ * Definition for singly-linked list.
+ * struct ListNode {
+ *     int val;
+ *     ListNode *next;
+ *     ListNode() : val(0), next(nullptr) {}
+ *     ListNode(int x) : val(x), next(nullptr) {}
+ *     ListNode(int x, ListNode *next) : val(x), next(next) {}
+ * };
+ */
+#include <stack>
+class Solution {
+public:
+    ListNode* reverseList(ListNode* head) {
+        if (head == nullptr) {
+            return head;
+        }
+
+        ListNode* node = head;
+        stack<int> vals;
+        while (node) {
+            vals.push(node->val);
+            node = node->next;
+        }
+
+        ListNode* reverse_node = new ListNode();
+        node = reverse_node;
```

Comment:
ヒープにメモリを確保する必要がなさそうです。今のコードだとreverse_nodeはメモリリークしますね

```suggestion
        ListNode reverse_node = ListNode();
        node = &reverse_node;
```

Reply:
参考：
automatic storage duration (stack) vs dynamic storage duration (heap)
https://en.cppreference.com/w/cpp/language/storage_duration
https://en.wikipedia.org/wiki/Automatic_variable

Reply:
@fhiyo @liquo-rice 
指摘&&資料有難うございます。
メモリに対する意識がまだ低いようです。

**質問です。**
new演算子を使う場合は、deleteをつける対応でも良いのでしょうか？
もしくはheapメモリを使う必要のない場合はできるだけstackメモリを使うものでしょうか？

Reply:
> もしくはheapメモリを使う必要のない場合はできるだけstackメモリを使うものでしょうか？

自分は基本こっちだと思っています。ヒープにメモリを確保するのは単にスタックポインタをずらすだけのローカル変数の定義よりもコストがかかるので。あとnewするとどこでそのメモリを解放するのかを気にするので読み手の負荷が高くなる気はします